File: lab6-openmp/lab6-openmp_f90/omp_getEnvInfo.f90
! -*- Mode: Fortran90; -*-
!=================================================================

program GetInfo
  !-----------------------------------------------------------------
  ! FILE: omp_getEnvInfo.f90
  ! DESCRIPTION:
  !   OpenMP Example - Get Environment Information
  !   The master thread queries and prints selected environment
  !   information.
  ! AUTHOR: Blaise Barney  7/06
  ! Upgraded to F90, free-source form: Daniel R. Reynolds, 2/2009
  !-----------------------------------------------------------------
  !======= Inclusions ===========

  !======= Declarations =========
  implicit none

  integer :: NThreads, TID, Procs, MaxT
  integer, external :: omp_get_num_threads, omp_get_thread_num, &
       omp_get_num_procs, omp_get_max_threads, &
       omp_in_parallel, omp_get_dynamic, omp_get_nested
  logical :: InPar, Dyn, Nested

  !======= Internals ============

  ! Start parallel region
  !$omp parallel private(TID)
  
  ! Obtain thread number
  TID = omp_get_thread_num()
  
  ! Only master thread does this
  if (TID == 0) then
     
     print *, 'Thread',tid,'getting environment information'
     
     ! Get environment information
     Procs    = omp_get_num_procs() 
     NThreads = omp_get_num_threads()
     MaxT     = omp_get_max_threads()
     InPar    = omp_in_parallel()
     Dyn      = omp_get_dynamic()
     Nested   = omp_get_nested()

     ! Print environment information
     print *, 'Number of processors = ', Procs
     print *, 'Number of threads = ', NThreads
     print *, 'Max threads = ', MaxT
     print *, 'In parallel? = ', InPar
     print *, 'Dynamic threads enabled? = ', Dyn
     print *, 'Nested parallelism supported? = ', Nested

  end if
  
  ! Done
  !$omp end parallel

end program GetInfo
!=================================================================

File: lab6-openmp/lab6-openmp_c/omp_getEnvInfo.c
/* FILE: omp_getEnvInfo.c
   DESCRIPTION:
      OpenMP Example - Get Environment Information - C Version
      The master thread queries and prints selected environment information.
   AUTHOR: Blaise Barney  7/06
   UPDATED: Daniel R. Reynolds, 1/13/2013 */

#include <omp.h>
#include <stdio.h>
#include <stdlib.h>

int main (int argc, char *argv[]) {

  /* local variables */
  int nthreads, tid, procs, maxt, inpar, dynamic, nested;

  /* Start parallel region */
# pragma omp parallel private(nthreads, tid)
  {

    /* Obtain thread number */
    tid = omp_get_thread_num();

    /* Only master thread does this */
    if (tid == 0) {
      printf("Thread %d getting environment info...\n", tid);

      /* Get environment information */
      procs    = omp_get_num_procs();
      nthreads = omp_get_num_threads();
      maxt     = omp_get_max_threads();
      inpar    = omp_in_parallel();
      dynamic  = omp_get_dynamic();
      nested   = omp_get_nested();

      /* Print environment information */
      printf("Number of processors = %d\n", procs);
      printf("Number of threads = %d\n", nthreads);
      printf("Max threads = %d\n", maxt);
      printf("In parallel? = %d\n", inpar);
      printf("Dynamic threads enabled? = %d\n", dynamic);
      printf("Nested parallelism supported? = %d\n", nested);

    }

  }  /* end parallel region */

  return 0;
}  /* end main */

